File: miko_assignment.py
import matplotlib.pyplot as plt
import math
import random

fig, ax = plt.subplots()
##global cumulative_distance = 0
def populateObstacles(obstacleCenters, obsRadius, obstacles):
    obs1 = plt.Circle(obstacleCenters[0], obsRadius[0], color='r', fill=False)
    obs2 = plt.Circle(obstacleCenters[1], obsRadius[1], color='r', fill=False)
    obs3 = plt.Circle(obstacleCenters[2], obsRadius[2], color='r', fill=False)
    obs4 = plt.Circle(obstacleCenters[3], obsRadius[3], color='r', fill=False)
    obs5 = plt.Circle(obstacleCenters[4], obsRadius[4], color='r', fill=False)

    ax.add_patch(obs1)
    ax.add_patch(obs2)
    ax.add_patch(obs3)
    ax.add_patch(obs4)
    ax.add_patch(obs5)

    for i in obstacleCenters:
        obstacles.append(i)
    
    for i in range(len(obstacleCenters)):
        x = obstacleCenters[i][0]
        y = obstacleCenters[i][1]
        r = obsRadius[i]
        for j in range(r+1):
            obstacles.append((x + j, y))
            obstacles.append((x - j, y))

            obstacles.append((x, y + j))
            obstacles.append((x, y - j))

            obstacles.append((x + r, y + j))
            obstacles.append((x + r, y - j))
            obstacles.append((x - r, y + j))
            obstacles.append((x - r, y - j))

            obstacles.append((x - j, y + r))
            obstacles.append((x + j, y + r))
            obstacles.append((x + j, y - r))
            obstacles.append((x - j, y - r))

            #obstacles.append((x - j, y + j))
            #obstacles.append((x - j, y + j))

            obstacles.append((x + j, y + j))
            obstacles.append((x - j, y - j))
            obstacles.append((x + j, y - j))
            obstacles.append((x - j, y + j))   
                 

def getLineCoefs(p1, p2):
    if p1[0] != p2[0] and p1[1] != p2[1]:
        a = p2[1] - p1[1]
        b = p1[0] - p2[0]
        c = a*(p1[0]) + b*(p1[1])
        return [a,b,c]
    return 0

def checkCollision(line, center, radius):
    a = line[0]
    b = line[1]
    c = line[2]
    x = center[0]
    y = center[1]
    dist = ((abs(a * x + b * y + c)) /
            math.sqrt(a * a + b * b))
    if radius > dist:
        return True
    return False

#def plotRobot(pos):
#    robotRadius = 0.2
#    robot = plt.Circle(pos, robotRadius, color='blue', fill=True)
#    ax.add_patch(robot)

def getDistance(p1, p2):
    return math.dist(p1, p2)

def updateRobotPose(obstacleCenters, obsRadius, obstacles, initialPose):
    outPoses = []
    for m in range(10):
        plt.scatter(initialPose[0], initialPose[1], c='b')
        possibleStates = []
        stepSize = 5
        updatedState = 0
        for i in range(-10, 11, 1):
            for j in range(-10, 11, 1):
                if (i,j) not in obstacles:
                    possibleStates.append((i,j))
        
        possibleStates = set(possibleStates)
        obstacles = set(obstacles)
        possibleStatesTemp = possibleStates.copy()
        obstaclesTemp = obstacles.copy()

        for i in obstacles:
            if getDistance(i, initialPose) > stepSize:
                obstaclesTemp.remove(i)
        
        for i in possibleStatesTemp:
            if getDistance(i, initialPose) != stepSize:
                possibleStates.remove(i)

        possibleStatesTemp = possibleStates.copy()
        for i in possibleStatesTemp:
            for j in obstaclesTemp:
                if i[0] == j[0] or i[1] == j[1]:
                    possibleStates.remove(i)
                    break
        
        possibleStatesTemp = possibleStates.copy()
        for i in possibleStatesTemp:
            if i[0] == initialPose[0] or i[1] == initialPose[1]:
                break
            else:
                line = getLineCoefs(i, initialPose)
                for j in range(len(obstacleCenters)):
                    if getDistance(obstacleCenters[j], initialPose) < stepSize:
                        if checkCollision(line, obstacleCenters[j], obsRadius[j]):
                            possibleStates.remove(i)
                            break
        possibleStates = list(possibleStates)
        updatedState = random.choice(possibleStates)
        initialPose = updatedState
        ##cumulative_distance = cumulative_distance + getDistance(updatedState, initialPose)

        outPoses.append(updatedState)
    return outPoses

if __name__ == "__main__":
    obstacleCenters = [(0,0), (-3, 5), (4, 6), (6, -7), (-8, -7)]
    obsRadius = [1, 2, 2, 1, 2]
    obstacles = []
    moving_cumulative_distance_list = []
    populateObstacles(obstacleCenters, obsRadius, obstacles)
    initialPose = (-10, -10)
    poses = updateRobotPose(obstacleCenters, obsRadius, obstacles, initialPose)
    print(poses)
    distance_between_poses = [getDistance(poses[i+1],poses[i]) for i in range(len(poses)-1)]
    print(distance_between_poses)
    print(sum(distance_between_poses))
    
    moving_cumulative_distance = 0
 
    for i in range(0, len(poses)-1):
        moving_cumulative_distance = moving_cumulative_distance + getDistance(poses[i+1],poses[i]) 
        ##print("moving_cumulative_distance: ", moving_cumulative_distance)
        moving_cumulative_distance_list.append(moving_cumulative_distance)
        ##print("moving_cumulative_distance_list: ", moving_cumulative_distance_list)


    for i in range(0, len(poses)-1):
        ##print("waypoint:",i,poses[i],"parent:",poses[i-1],"moving_cumulative_distance",moving_cumulative_distance_list[i])

        data_structure ={"waypoint": poses[i], "parent": poses[i-1],"moving_cumulative_distance": moving_cumulative_distance_list[i]}
        print (data_structure)
    for i in poses:
        
        plt.scatter(i[0], i[1], c='blue')
    ax.plot()
    plt.show()

File: miko_assignment.cpp
#include <iostream>
#include <cmath>
using namespace std;
#include <vector>
#include <algorithm>

#include <iterator> 
#include <map> 

// struct ObstacleCenters
// {
//     int x, y;
// };
struct Points
{
    int x, y;
};

// int a,b,c,x,y,dist;

bool checkCollision(int line[], Points& center, int& radius)
{
    
    int a,b,c,x,y,dist;
    a = line[0];
    b = line[1];
    c = line[2];
    x = center.x;
    y = center.y;
    dist = ((abs(a * x + b * y + c)) /
            sqrt(a * a + b * b));
    if (radius > dist)
        return true;
    return false; 

}


float distance(int x1, int y1, int x2, int y2)
{
    // Calculating distance
    return sqrt(pow(x2 - x1, 2) + pow(y2 - y1, 2) * 1.0);
}



int main()
{
// vector<ObstacleCenters> ocs = {{0,0},{-3, 5},{4, 6},{6, -7},{-8, -7}};
vector<int> moving_cumulative_distance_list;
vector<int> distance_between_poses;
int moving_cumulative_distance;
map<string, Points> data_structure;


int radius[] = { 1, 2, 2, 1, 2};
vector<Points> obstacleCenters ={{0,0},{-3, 5},{4, 6},{6, -7},{-8, -7}};

vector<Points> obstacles;
Points initialPose{-10,-10};
int x,y,r;
vector<Points>poses;
srand((unsigned int)time(NULL));


/////////////////////////populateObstacles/////////////
// for (auto i : obstacleCenters)
//     // obstacles.push_back(i);
//     cout<<i<<endl;
// cout << '\n'; 
    
for (int i = 0; i <obstacleCenters.size() ; i++) {
    int x = obstacleCenters[i].x;
    int y = obstacleCenters[i].y;
    int r = radius[i];
    
    for (int j = 0; j < (r+1); j++) {
        obstacles.push_back({x + j, y});
        obstacles.push_back({x - j, y});
        
        obstacles.push_back({x, y + j});
        obstacles.push_back({x, y - j});
        
        obstacles.push_back({x + r, y + j});
        obstacles.push_back({x + r, y - j});
        obstacles.push_back({x - r, y + j});
        obstacles.push_back({x - r, y - j});

        obstacles.push_back({x - j, y + r});
        obstacles.push_back({x + j, y + r});
        obstacles.push_back({x + j, y - r});
        obstacles.push_back({x - j, y - r});

        obstacles.push_back({x + j, y + j});
        obstacles.push_back({x - j, y - j});
        obstacles.push_back({x + j, y - j});
        obstacles.push_back({x - j, y + r});
    }
    
    
  } 
    
// cout<<"size of obstacles "<<obstacles.size() <<endl; 
// for (auto i : obstacles)

//     cout <<" ox:"<< i.x <<" oy:"<<i.y<<endl;   

    
int count=0;
      
//  /////////////////////////update robot pose/////////////
vector<Points> outPoses;

for (int i = 0; i < 10; i++) {
  vector<Points> possibleStates;
  int stepSize = 1;
  Points updatedState;
  int line[3];

  vector<Points> possibleStatesTemp;
  vector<Points> obstaclesTemp;
  int a,b,c;
  
  //int updatedState = 0;
  for (int i = -10; i < 11; i++) {
     for (int j = -10; j < 11; j++) {
        // if i,j is prenent in obstacles
            auto pred = [i,j](const Points & item) {
            return item.x == i,item.y == j;
            };
             //cout<<(std::find_if(std::begin(obstacles), std::end(obstacles), pred) != std::end(obstacles))<<endl;
            if (std::find_if(std::begin(obstacles), std::end(obstacles), pred) != std::end(obstacles)){
                possibleStates.push_back({i,j});
            }
            
         }
         
     }  
//   for (auto i : possibleStates)
//      cout <<"x:"<< i.x <<"y:"<<i.y<<endl;  
   possibleStatesTemp = possibleStates;
   obstaclesTemp = obstacles;   
   //cout << obstacles.size()<<endl;
   for (auto i : obstacles){
       if (distance(i.x , i.y, initialPose.x, initialPose.y) > stepSize)
       {
       obstaclesTemp.erase(std::remove_if(obstaclesTemp.begin(), obstaclesTemp.end(), [&](Points const & item) {
            return item.x == i.x, item.y == i.y;
        }),obstaclesTemp.end());
           
       }
   }
    //  cout <<"size of possibleStates before:  "<< possibleStates.size() <<endl; 
   
  for (auto i : possibleStatesTemp){
      if (distance(i.x , i.y, initialPose.x, initialPose.y) > stepSize)
      {
      possibleStates.erase(std::remove_if(possibleStates.begin(), possibleStates.end(), [&](Points const & item) {
            return item.x == i.x,item.y == i.y;
        }),possibleStates.end());
           
      }
  }
   
    //  cout <<"size of possibleStates after:  "<< possibleStates.size() <<endl; 

   possibleStatesTemp = possibleStates;
   for (auto i : possibleStatesTemp){
       for (auto j : obstaclesTemp){
    //  cout <<"x:"<< i.x <<"y:"<<i.y<<endl; 
    //  cout <<"x:"<< j.x <<"y:"<<j.y<<endl; 
       
       
       if (i.x == j.x , i.y ==j.y )
       {
        possibleStates.erase(std::remove_if(possibleStates.begin(), possibleStates.end(), [&](Points const & item) {
            return item.x == i.x && item.y == i.y;
        }),possibleStates.end());
        
        break;   
       }
       }
   }

//   for (auto i : possibleStates)                  ///no element in possibleStates?
//      cout <<"x:"<< i.x <<"y:"<<i.y<<endl; 


   possibleStatesTemp = possibleStates;
   ///////////////////////////
   for (auto i : possibleStatesTemp){
       if (i.x == initialPose.x || i.y ==initialPose.y )
       {
           break;
       }   
       else {    

       if ((i.x != initialPose.x) && (i.y != initialPose.y)) 
       {
        a = initialPose.y - i.y;
        b = i.x  - initialPose.x; 
        c = a*(i.x) + b*(i.y);
        int line[] = {a,b,c};    
        
        int z=0;   
        for (auto j : obstacleCenters){
        if (distance(j.x,j.y,initialPose.x,initialPose.y) < stepSize){
            
            if (checkCollision(line, j, radius[z]))  {
       
                possibleStates.erase(std::remove_if(possibleStates.begin(), possibleStates.end(), [&](Points const & item) {
                    return item.x == i.x,item.y == i.y;
                }),possibleStates.end());
        
                break; 
   
            }
        z=z+1;
          }    
        }
       }

        }


      
    }
   
//   possibleStates = list(possibleStates)?

   updatedState = possibleStates[rand() % possibleStates.size()];
//   cout <<"x:"<< updatedState.x <<"y:"<<updatedState.y<<endl;  

   initialPose = updatedState;
   outPoses.push_back(updatedState);
  }
 for (auto i : outPoses)
    cout <<"x:"<< i.x <<"y:"<<i.y<<endl;     
 poses = outPoses;
 
 
 
 
 moving_cumulative_distance = 0;
 
 for (int i=0; i<poses.size();i++)
 {
     int d = distance(poses[i].x , poses[i].y, poses[i+1].x , poses[i+1].y);
     distance_between_poses.push_back(d); 
     
     moving_cumulative_distance = moving_cumulative_distance + distance(poses[i].x , poses[i].y, poses[i+1].x , poses[i+1].y);
     moving_cumulative_distance_list.push_back(moving_cumulative_distance);
     

 }
 
 
//  moving_cumulative_distance[0];
 
// int f =0; 
// for (auto i : poses)
//     //  data_structure = {{"waypoint", i}, {"parent", i},{"moving_cumulative_distance", moving_cumulative_distance_list[f]}};
//      data_structure = {{"waypoint", i}, {"parent", i}};

//      f = f + 1;
return 0;
}



























